test-cursors/Test/Database/LMDB/Simple/Cursor/Lockstep/Mock.hs

```diff
@@ -0,0 +1,360 @@
+  , mCursorGetSetKey
+  , mCursorGetSetRange
+    -- * Mock cursor operations: put
+  , mCursorPut
+    -- * Mock cursor operations: delete
+  , mCursorDel
+    -- * Utility: viewing cursor positions
+  , CPosView (..)
+  , viewCPos
+    -- * Convert observable errors to mock errors
+  , fromLMDBError
+  ) where
+
+import           Control.Monad
+import           Control.Monad.Error.Class
+import           Control.Monad.Except
+import           Control.Monad.Identity
+import           Control.Monad.State.Class
+import           Control.Monad.State.Strict
+import           Data.Map.Strict             (Map)
+import qualified Data.Map.Strict             as Map
+
+import           Database.LMDB.Raw
+import           Database.LMDB.Simple.Cursor
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+data Mock k v = Mock {
+    -- | The key-value pairs that are stored in the database.
+    store :: Map k v
+    -- | Position of the cursor in the key-value store. @'Nothing'@ signals that
+    -- the cursor position is a null pointer.
+  , cpos  :: Maybe k
+  }
+  deriving (Show, Eq)
+
+emptyMock :: Mock k v
+emptyMock = Mock {
+    store = Map.empty
+  , cpos  = Nothing
+  }
+
+-- | Simple enumeration type for the three types of cursor oprations: get, put
+-- and delete.
+data COp = Get | Put | Del
+  deriving (Show, Eq)
+
+data Err =
+    -- | An @'MDB_cursor_op'@ is not supported for a get operation.
+    ErrCursorOpNotSupported MDB_cursor_op
+    -- | A @'CPutFlag'@ is not supported for a put operation.
+  | ErrPutFlagNotSupported CPutFlag
+    -- | A @'CDelFlag'@ is not supported for a delete operation.
+  | ErrDelFlagNotSupported CDelFlag
+    -- | A cursor operation got an invalid argument, like a cursor position that
+    -- is a null pointer.
+  | ErrInvalidArgument COp
+    -- | A specific key-value pair was not found in the database.
+    --
+    -- Example: deletes will delete the first key-value pair in a cursor
+    -- position that is greater than or equal to the current cursor position.
+    -- This means that if the cursor points to a key-value pair, then it will
+    -- delete it. If it does not point to a key-value pair, then we delete the
+    -- closest, successor key-value pair. However, if there no more successors,
+    -- then this error is thrown.
+  | ErrNotFound COp
+
+  deriving (Show, Eq)
+
+{-------------------------------------------------------------------------------
+  Mocked cursor monad
+-------------------------------------------------------------------------------}
+
+-- | Mocked cursor monad.
+newtype MC k v a = MC {unMC :: ExceptT Err (StateT (Mock k v) Identity) a}
+  deriving stock Functor
+  deriving newtype (Applicative, Monad, MonadState (Mock k v), MonadError Err)
+
+runMC :: MC k v a -> Mock k v -> (Either Err a, Mock k v)
+runMC mc m = runIdentity (runStateT (runExceptT (unMC mc)) m)
```

Comment:
Would it make sense to avoid parenthesis using point-free notation?

Reply:
Changed to:
```haskell
runMC :: MC k v a -> Mock k v -> (Either Err a, Mock k v)
runMC = runState . runExceptT . unMC
```